### PR TITLE
[Driver][SYCL] Do not pass along implied PVC specific values for AOT

### PR DESCRIPTION
**Pull request by mdtoguchi**
When compiling for AOT GPU, we passed along default device settings for PVC regardless of intended device.  Update this behavior to only pass along the PVC specific options to ocloc when the matching device (i.e. -device pvc) is also supplied.

The device can be specific with the typical device value (pvc) or via the -fsycl-targets=intel_gpu_pvc option setting or -device values that match with PVC that match a given version or HEX value.

The default values will continue to be passed along for JIT.

### CODE REVIEW

**Review by sarnex — APPROVED** [review]
Looks great, thanks a lot for the fast turnaround and sorry for making you fix what is basically my code,

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 1]
```diff
@@ -1415,15 +1495,21 @@ void SYCLToolChain::AddImpliedTargetArgs(const llvm::Triple &Triple,
...
-             Triple.getSubArch() == llvm::Triple::NoSubArch) {
-      BeArgs.push_back(Args.MakeArgString(RegAllocModeOptName + DeviceName +
-                                          ":" + BackendOptName));
+    // pvc:default on Windows and and pvc:auto otherwise when -device pvc is
+    // provided by the user.
+    ArgStringList TargArgs;
+    Args.AddAllArgValues(TargArgs, options::OPT_Xs, options::OPT_Xs_separate);
+    Args.AddAllArgValues(TargArgs, options::OPT_Xsycl_backend);
+    // Check for any -device settings.
+    if (hasPVCDevice(TargArgs) || Device == "pvc" || IsJIT) {
```
[the remark this answers is not in the document]
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp:1503]: makes sense.  I'll update.

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 2]
```diff
@@ -865,6 +865,83 @@ void SYCL::fpga::BackendCompiler::ConstructJob(
...
+  // when ranges or wildcards are used.
+  // Here we parse the targets, tokenizing via ','
+  SmallVector<StringRef> SplitArgs;
+  DeviceArg.split(SplitArgs, ",");
+  for (const auto &SingleArg : SplitArgs) {
+    StringRef OclocTarget;
+    // Handle shortened versions.
+    bool CheckShortVersion = true;
+    for (auto Char : SingleArg.str()) {
+      if (!(std::isdigit(Char) || Char == '.')) {
```
[the remark this answers is not in the document]
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp:922]: Readability is key - will update.

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (sarnex) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: Ah got it, thanks

**Review by sarnex — APPROVED** [review]

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 3]
```diff
@@ -865,6 +865,83 @@ void SYCL::fpga::BackendCompiler::ConstructJob(
...
+    }
+    if (DeviceSeen)
+      break;
+  }
+  if (DeviceArg.empty())
+    return false;
+
+  // Go through all of the arguments to '-device' and determine if any of these
+  // are pvc based.  We only match literal values and will not find a match
+  // when ranges or wildcards are used.
```
Comment (srividya-sundaram) [inline clang/lib/Driver/ToolChains/SYCL.cpp:913]: > We only match literal values and will not find a match when ranges or wildcards are used.

Is there a test case for this case?

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp:913]: No, but I will add a couple.

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp:913]: Tests added

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 4]
```diff
@@ -1415,15 +1494,21 @@ void SYCLToolChain::AddImpliedTargetArgs(const llvm::Triple &Triple,
...
-    StringRef BackendOptName = SYCL::gen::getGenGRFFlag("auto");
-    if (IsGen)
-      PerDeviceArgs.push_back({DeviceName, Args.MakeArgString(BackendOptName)});
-    else if (Triple.isSPIROrSPIRV() &&
-             Triple.getSubArch() == llvm::Triple::NoSubArch) {
-      BeArgs.push_back(Args.MakeArgString(RegAllocModeOptName + DeviceName +
-                                          ":" + BackendOptName));
+    // pvc:default on Windows and and pvc:auto otherwise when -device pvc is
+    // provided by the user.
+    ArgStringList TargArgs;
```
Comment (srividya-sundaram) [inline clang/lib/Driver/ToolChains/SYCL.cpp:1499]: nit: `ArgStringList TargArgs;` is declared 3 times at different places in this file. Is there a way to optimize them? Like adding them as a parameter in the function there are used?

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp:1499]: 2 uses in the file are similar (gathering `-X*` options), the other is not related.  The similar entries can be consolidated but use is dependent on various behaviors so there is never a guarantee that the populated list would be used.  I can look into consolidating but would like to do it in a separate PR.

**Review by srividya-sundaram — APPROVED** [review]

**Comment by mdtoguchi** [comment]
@intel/llvm-gatekeepers, this PR looks ready to merge - thanks!

**Pull request merged**